### DISCUSSION TITLE
Make save and restore states work as a stack

### DISCUSSION THREAD

**Discussion by jesi-rgb** [ManimCommunity/manim#2813] [Suggestions and Proposals]
Hello everyone! Thanks so much beforehand for your precious time. I already read the docs and saw that new features won't be tackled because of the major refactor, but I do believe it is beneficial to post it nevertheless so I don't forget about it. I myself can try to tackle this problem in the future.

## Description of proposed feature

Right now, as of version 0.15.1, save and restore states work with a single state. Code can be seen [here](https://docs.manim.community/en/stable/_modules/manim/mobject/mobject.html#Mobject.save_state). The way it currently works is just saving a copy of the current state of the `Mobject`. In case I called the `Mobject.save_state()` twice, the first one would be overwritten.

I found a bit of trouble working with this setup using `self.camera.frame.save_state()` because I had two different, kind of nested states: the default camera view, one zoomed in and one even more zoomed in. If I want to go from general to zoomed, then zoom a bit more, I will lose the general camera overview. There are ways around it, but pretty hacky.

I propose that this functionality works as a stack, as in classic OpenGL `push()` and `pop()` matrices. If I save a state, it will be saved in a LIFO array. I can save many times and then pop as needed.

## How can the new feature be used?

The way I imagine the interface to look is exactly the same as it works now. For any given mobject, we can do something like:

``` python
mob = Circle()
mob.save_state() # default state
self.play(mob.animate.scale(2)) # scaling 1
mob.save_state()
self.play(mob.animate.scale(2)) # scaling 2
self.play(mob.restore()) # return to scaling 1
self.play(mob.restore()) # return to default state
```

## Additional comments

I don't know if there are official ways to support this or if this can interfere negatively in how the interface works. As I've seen many times in other issues and PRs, keeping the usage as simple and as easy as possible is one of the main goals of the library.

I think that by the time a given user has the necessity to do what it is described here, it can be (arguably) safe to assume that they know about the `push()` and `pop()` mechanic.

In case they didn't, it would definitely be beneficial to know. But that is totally open for discussion.

Thanks again. Let me know if I missed something!

**Comment by behackl** [comment]
I like this idea!

**Comment by jesi-rgb** [comment]
That's great. Should I wait for a 1.0 version of Manim to implement it? What do you think?

**Comment by naveen521kk** [comment]
> Should I wait for a 1.0 version of Manim to implement it? What do you think?

You could make a PR if you like, now. You won't need to wait for the refactor to complete.
